src/optimal_image/search.rs

```diff
@@ -1,35 +1,127 @@
-use dataclients::ImageData;
+use dataclients::*;
+use image::*;
+use std::collections::HashMap;
+use std::error::Error;
 use std::path::Path;
+extern crate dssim;
+use EncodingConfig;
 use ImageConfig;
 use ImageFormat;
 use ImageScore;
 
+#[derive(Debug)]
 pub struct SearchOptions {
-    threshold: f64,
-    formats: Vec<ImageFormat>,
+    pub threshold: f64,
+    pub quality_range: (u8, u8),
+    pub formats: Vec<ImageFormat>,
 }
 
+pub type SearchResult = HashMap<ImageFormat, ImageConfig>;
+
 pub struct Search {
+    path: String,
     image_data: ImageData,
     options: SearchOptions,
+    result: Option<SearchResult>,
+    dssim: dssim::Dssim,
```

Comment:
I think we may not need `dssim` as member field for every `Search` instance

Reply:
I'm not sure about the cost of creating this instance, here I'm assuming that would be cheaper to reuse the same dssim context during a whole search instead of creating a new one for each comparison, but we can benchmark that.